### ISSUE TITLE
Nested volumes sometimes not mounted correctly

### ISSUE THREAD

**Issue by pascalgn** [kubernetes/kubernetes#71800]
**What happened**:
When running Jenkins in Kubernetes on Azure, we have two volumes, `/var/jenkins_home` (which is on a CIFS network share) and `/var/jenkins_home/workspaces` (which is from a persistent volume). When the Jenkins pod is starting, for example after the node has rebooted, it can happen that /var/jenkins_home gets mounted, but /var/jenkins_home/workspaces does not.

**What you expected to happen**:
The path /var/jenkins_home/workspaces should only get mounted _after_ /var/jenkins_home has been mounted. I suspect that there's some timing issue where sometimes /var/jenkins_home/workspaces gets mounted after /var/jenkins_home is being mounted (especially considering /var/jenkins_home is from a network share).

**How to reproduce it (as minimally and precisely as possible)**:
```yaml
apiVersion: apps/v1beta2
kind: StatefulSet
metadata:
  name: jenkins
spec:
  serviceName: jenkins
  selector:
    matchLabels:
      app: jenkins
  replicas: 1
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: jenkins
    spec:
      containers:
        - name: jenkins
          image: jenkins:2.154
          ports:
            - containerPort: 80
              name: jenkins-http
          volumeMounts:
            - mountPath: /var/jenkins_home
              name: jenkins-home
            - mountPath: /var/jenkins_home/cache
              name: jenkins-home-caches
            - mountPath: /var/jenkins_home/caches
              name: jenkins-home-caches
            - mountPath: /var/jenkins_home/tools
              name: jenkins-home-tools
            - mountPath: /var/jenkins_home/updates
              name: jenkins-home-updates
            - mountPath: /var/jenkins_home/workspace
              name: jenkins-home-workspace
      volumes:
        - name: jenkins-home
          azureFile:
            secretName: storage-secret
            shareName: jenkins-home
            readOnly: false
        - name: jenkins-home-cache
          emptyDir: {}
        - name: jenkins-home-caches
          emptyDir: {}
        - name: jenkins-home-updates
          emptyDir: {}
  volumeClaimTemplates:
    - metadata:
        name: jenkins-home-workspace
        annotations:
          volume.beta.kubernetes.io/storage-class: managed-premium
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 75Gi
    - metadata:
        name: jenkins-home-tools
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
---
apiVersion: v1
kind: Service
metadata:
  name: jenkins
  labels:
    app: jenkins
spec:
  type: LoadBalancer
  ports:
    - port: 80
      name: service-http
  selector:
    app: jenkins
```
I have omitted the parts that are (probably) irrelevant.

Note that there is also a second mount, `/var/jenkins_home/tools`, which did get mounted even when `/var/jenkins_home/workspaces` was not mounted.

**Anything else we need to know?**:
The issue does not occur all the time. It can be resolved by connecting to the pod and killing Jenkins, so that the container restarts. After the container restart, all volumes are mounted correctly.

**Environment**:
- Kubernetes version (use `kubectl version`): Client: v1.12.2, Server: v1.11.1
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): Debian GNU/Linux 9 (stretch)
- Kernel (e.g. `uname -a`): Linux jenkins-0 4.15.0-1032-azure #33~16.04.1-Ubuntu SMP Fri Nov 9 22:36:11 UTC 2018 x86_64 GNU/Linux
- Install tools: acs-engine
- Others: -

/kind bug

**Comment by andyzhangx** [comment]
@pascalgn since `jenkins-home` is an existing azure file and `jenkins-home-workspace` is a new disk, usually mounting `jenkins-home` should be faster. While all mount points should not have mounting time sequence.

**Comment by pascalgn** [comment]
I already suspected this might not necessarily be a Kubernetes issue, at least not exclusively. I searched a bit more and found:

>The issue is that from time to time the "/home/bigdir" folder will be empty, even though mtab think that the share is still mounted. Only thing that works is by unmounting, and then (re)mounting the bigdir share.
> _[NFS mount mounted inside another NFS mount disappears randomly
](https://serverfault.com/questions/403550/nfs-mount-mounted-inside-another-nfs-mount-disappears-randomly)_

A similar entry is _[NFS mounts depending on other NFS mount fail to mount at boot.
](https://serverfault.com/questions/763997/nfs-mounts-depending-on-other-nfs-mount-fail-to-mount-at-boot)_ There are workarounds mentioned, like using autofs or netfs (not sure if they also support CIFS).

So this seems to be a general issue when doing nested mounting, that is mounting network shares inside other mounted network shares.

**I think it would make sense to do one or more of the following:**
1. Update [the documentation](https://kubernetes.io/docs/concepts/storage/volumes/) to warn users that network volumes should not be nested, i.e. a network volume should not be mounted inside another network volume mount point
2. Issue a warning event when a network volume is mounted inside another network volume
3. Implement some kind of retry logic, so when the parent directory gets remounted (for example due to network issues), the child directories should also get remounted

WDYT?
